### ISSUE TITLE
Weights not initialized warning in reward function

### ISSUE THREAD

**Issue by Lamply** [OptimalScale/LMFlow#561]
I tried RAFT alignment following the example at https://optimalscale.github.io/LMFlow/examples/raft.html and used following command to obtain the merge reward model:
```sh
python examples/merge_lora.py \
    --model_name_or_path output_models/sft \
    --lora_model_path output_models/rm \
    --output_model_path output_models/merge_rm
```
 
However, I encountered a warning while performing RAFT alignment:

> Some weights of GPTNeoForSequenceClassification were not initialized from the model checkpoint at output_models/merge_rm and are newly initialized: ['score.weight']
> You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.

It seems that the last layer of the reward model was not loaded properly. Is this expected behavior? How should I test or evaluate the reward function?

**Comment by WeiXiongUST** [comment]
Hi! Thanks for your interest!

Just to clarify. Do you meet this warning when you load the merged reward model when running raft?

**Comment by WeiXiongUST** [comment]
You can also try out the following code. 

```python
import torch
from peft import PeftConfig, PeftModel
from transformers import AutoConfig, AutoTokenizer, HfArgumentParser, AutoModelForSequenceClassification

model = AutoModelForSequenceClassification.from_pretrained(base_model_name, num_labels=1, torch_dtype=torch.bfloat16)
model_lora = PeftModel.from_pretrained(model, adapter_model_name)
model_lora.merge_and_unload()
model.save_pretrained(output_name)

**Comment by WeiXiongUST** [comment]
To test whether the reward model is merged appropriately, you can load the rm dataset and use your reward model to test the accuracy. For instance,

```python
dataset = load_dataset("json", data_files="/home/usr_name/LMFlow/data/hh_rlhf/rm/the_name_of_rm_data.json", split="train", field="instances")
responses_j = dataset['positive']
responses_k = dataset['negative']
cnt = 0
rm_tokenizer = AutoTokenizer.from_pretrained(reward_model_or_path)
hf_pipe = pipeline(
    "sentiment-analysis",
    model=reward_args.reward_model_or_path,
    device="gpu",
    tokenizer=rm_tokenizer,
    model_kwargs={"torch_dtype": torch.bfloat16}
)
for i in range(len(responses_j)):
    r_j = responses_j[i]
    r_k = responses_k[i]
    texts_for_rewards = [r_j, r_k]
    pipe_outputs = hf_pipe(texts_for_rewards, **pipe_kwargs)
    rewards = [output[0]["score"] for output in pipe_outputs]
    if rewards[0] >= rewards[1]:
        cnt += 1

print(cnt / len(responses_j))

**Comment by Lamply** [comment]
> Hi! Thanks for your interest!
> 
> Just to clarify. Do you meet this warning when you load the merged reward model when running raft?

Yes, both setting `pipeline` or using `AutoModel.get_model(reward_model_args)` will meet this issue.


> You can also try out the following code.
> 
> ```python
> import torch
> from peft import PeftConfig, PeftModel
> from transformers import AutoConfig, AutoTokenizer, HfArgumentParser, AutoModelForSequenceClassification
> 
> model = AutoModelForSequenceClassification.from_pretrained(base_model_name, num_labels=1, torch_dtype=torch.bfloat16)
> model_lora = PeftModel.from_pretrained(model, adapter_model_name)
> model_lora.merge_and_unload()
> model.save_pretrained(output_name)
> ```
I found that the model used in this approach is the `GPTNeoForSequenceClassification` model, which achieves an accuracy of about 71.8% when tested on the code you mentioned (only the first 1648 training data were run). 

However, the examples uses `GPTNeoForCausalLM`, where the `score` layer is discarded and replaced by the `lm_head` layer, which is also ignored when saving. 

Therefore, the models used in `merge_lora.py` and `raft_align.py` do not include the last layer of the reward model, and the output results will be random. Actually, I repeated a few times of `merge_lora.py` and tested them, and the accuracy fluctuated around 53%, instead of a fixed value.

**Comment by Lamply** [comment]
For more information:
### Method 1
```python
model = AutoModel.get_model(model_args)
model.merge_lora_weights()
model1 = model.get_backend_model()
model1
```
```
> PeftModelForCausalLM(
  (base_model): LoraModel(
    (model): GPTNeoForCausalLM(
      (transformer): GPTNeoModel(
        (wte): Embedding(50257, 2560)
        (wpe): Embedding(2048, 2560)
        (drop): Dropout(p=0.0, inplace=False)
        (h): ModuleList(
          (0-31): 32 x GPTNeoBlock(
            (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
            (attn): GPTNeoAttention(
              (attention): GPTNeoSelfAttention(
                (attn_dropout): Dropout(p=0.0, inplace=False)
                (resid_dropout): Dropout(p=0.0, inplace=False)
                (k_proj): Linear(in_features=2560, out_features=2560, bias=False)
                (v_proj): Linear(in_features=2560, out_features=2560, bias=False)
                (q_proj): Linear(in_features=2560, out_features=2560, bias=False)
                (out_proj): Linear(in_features=2560, out_features=2560, bias=True)
              )
            )
            (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
            (mlp): GPTNeoMLP(
              (c_fc): Linear(in_features=2560, out_features=10240, bias=True)
              (c_proj): Linear(in_features=10240, out_features=2560, bias=True)
              (act): NewGELUActivation()
              (dropout): Dropout(p=0.0, inplace=False)
            )
          )
        )
        (ln_f): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
      )
      (lm_head): Linear(in_features=2560, out_features=50257, bias=False)
    )
  )
)
```
### Method 2
```python
model2 = AutoModelForSequenceClassification.from_pretrained('output_models/finetune', num_labels=1, torch_dtype=torch.bfloat16)
model_lora = PeftModel.from_pretrained(model2, 'output_models/rm')
model_lora.merge_and_unload()
model2
```
```
> GPTNeoForSequenceClassification(
  (transformer): GPTNeoModel(
    (wte): Embedding(50257, 2560)
    (wpe): Embedding(2048, 2560)
    (drop): Dropout(p=0.0, inplace=False)
    (h): ModuleList(
      (0-31): 32 x GPTNeoBlock(
        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
        (attn): GPTNeoAttention(
          (attention): GPTNeoSelfAttention(
            (attn_dropout): Dropout(p=0.0, inplace=False)
            (resid_dropout): Dropout(p=0.0, inplace=False)
            (k_proj): Linear(in_features=2560, out_features=2560, bias=False)
            (v_proj): Linear(in_features=2560, out_features=2560, bias=False)
            (q_proj): Linear(in_features=2560, out_features=2560, bias=False)
            (out_proj): Linear(in_features=2560, out_features=2560, bias=True)
          )
        )
        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTNeoMLP(
          (c_fc): Linear(in_features=2560, out_features=10240, bias=True)
          (c_proj): Linear(in_features=10240, out_features=2560, bias=True)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.0, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
  )
  (score): Linear(in_features=2560, out_features=1, bias=False)
)
```

**Comment by WeiXiongUST** [comment]
We should use AutoModelForSequenceClassification as we train the rm by loading the GPTNeoForCausalLM as GPTNeoForSequenceClassification.  I think the accuracy 71.8% is normal for a 2.7b rm.

**Comment by shizhediao** [comment]
This issue has been marked as stale because it has not had recent activity. If you think this still needs to be addressed please feel free to reopen this issue. Thanks
